### PR TITLE
[rclcpp] Add class Waitable

### LINKED ISSUE #588: Add thing which adds things to wait set
## Feature request

This is a request for a feature to allow waiting on action servers and action clients. Action servers and Action clients are to be implemented in libraries on top of the core `rcl` and `rclcpp` features in `rcl_action` and `rclcpp_action`. Adding them to the wait set in `rcl` is covered by ros2/rcl#335.

### Feature description
This feature would allow an executor to wait on action servers and action clients which have servers/timers/clients/subscribers created in C. Since the individual entities don't have `rclcpp::<Entity>Base` types they can't be used by the existing callback group and executor APIs.

With #519 in mind it could make sense to change callback groups to be the class that adds things to the wait set, but I don't think there is time before the Crystal release. Instead this feature is proposing to add a new class to be used along with callback groups.

### Implementation considerations

Assuming class called `Waitable` with virtual APIs
* `virtual get_number_of_entities(size_t * num_subscriptions, ...)`
    * Called by executor when determining how big to allocate the wait set
* `virtual void add_to_wait_set(rcl_wait_set_t *)`
    * Called by executor when populating the wait set
    * Because `get_number_of_entities()` and `add_to_wait_set()` are separate calls, it must be assumed that the number of entities in a `Waitable` are constant between the two calls. For actions this assumption holds because the number of waitable entities is constant for the life of the server and client.
* `virtual bool can_execute(rcl_wait_set_t *)`
    * Return true if there is something that can be executed in the waitset.
* `virtual void execute();`
    * It's assumed the waitable will set some state within itself during `can_execute()` so that it does the right thing here.

Using `Waitable` requires:
* Adding a `Waitable::SharedPtr` field to [`AnyExecutable`](https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/any_executable.hpp)
* [`Executor::execute_any_executable()`](https://github.com/ros2/rclcpp/blob/master/rclcpp/src/rclcpp/executor.cpp#L264) needs to call `Waitable::execute()`
* Adding API `add_waitable(Waitable::SharedPtr)` to `Node`, possibly via `NodeWaitableInterface` (and an API to remove them)
* [`MemoryStrategy`](https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/memory_strategy.hpp) needs api `number_of_ready_waitables()`
* [`AllocatorMemoryStrategy::add_handles_to_wait_set()`](https://github.com/ros2/rclcpp/blob/c8f3fd3b0e424f4154dd7aee80fb26677ee852da/rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp#L183) needs to give `Waitable` a chance to add things
* `MemoryStrategy` needs api `get_next_waitable()` to be used by [`Executor::get_next_executable`](https://github.com/ros2/rclcpp/blob/master/rclcpp/src/rclcpp/executor.cpp#L530)

Requires ros2/rcl#335 so `Executor` and `Waitable` can know which entities belong to which.

### PR DESCRIPTION
**Pull request by jacobperron**
Provides a virtual API for interacting with wait sets.

Closes #588

### CODE REVIEW

File: rclcpp/include/rclcpp/waitable.hpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+  // TODO(jacobperron): smart pointer?
+  /// Add the Waitable to a wait set.
+  /**
+   * \param[in] wait_set A handle to the wait set to add the Waitable to.
+   * \return `true` if the Waitable is added successfully, `false` otherwise.
+   */
+  virtual bool add_to_wait_set(rcl_wait_set_t * wait_set) = 0;
+
+  // TODO(jacobperron): Rename to `is_ready()`?
```
Comment (sloretz) [inline rclcpp/include/rclcpp/waitable.hpp]: +1

File: rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp [hunk 2]
```diff
@@ -87,30 +87,42 @@ class AllocatorMemoryStrategy : public memory_strategy::MemoryStrategy
...
         client_handles_[i].reset();
       }
     }
-    for (size_t i = 0; i < wait_set->size_of_timers; ++i) {
+    for (size_t i = 0; i < timer_handles_.size(); ++i) {
       if (!wait_set->timers[i]) {
         timer_handles_[i].reset();
       }
     }
+    for (size_t i = 0; i < timer_handles_.size(); ++i) {
```
[the remark this answers is not in the document]
Comment (jacobperron) [reply rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp]: Copy-pasta :grimacing:. Fixed in a4a5493.

**Comment by sloretz** [comment]
Should add `NodeWaitables` to lifecycle node as well

**Comment by jacobperron** [comment]
> Should add NodeWaitables to lifecycle node as well

Thanks, LGTM.

**Comment by jacobperron** [comment]
With tests:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/5627/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2249/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/4648/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/5508/)

**Comment by sloretz** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/5631/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2251/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/4651/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/5511/)

**Comment by hidmic** [comment]
I *believe* Windows is failing because those `RCLCPP_PUBLIC` macros in ros2/system_tests#314 shouldn't be there. I'll take them out and re-trigger the job (if CI is there still).

**Comment by jacobperron** [comment]
Windows: [Build Status](https://ci.ros2.org/job/ci_windows/)

**Pull request merged**